registry/memory/memory.go

```diff
@@ -74,91 +118,152 @@ func (m *Registry) Init(opts ...registry.Option) error {
+		for version, r := range record {
+			if _, ok := m.records[name][version]; !ok {
+				m.records[name][version] = r
+				continue
+			}
+		}
 	}
-	m.Unlock()
+
 	return nil
 }
 
 func (m *Registry) Options() registry.Options {
 	return m.options
 }
 
-func (m *Registry) GetService(name string) ([]*registry.Service, error) {
-	m.RLock()
-	service, ok := m.Services[name]
-	m.RUnlock()
-	if !ok {
-		return nil, registry.ErrNotFound
+func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption) error {
+	m.Lock()
+	defer m.Unlock()
+
+	var options registry.RegisterOptions
+	for _, o := range opts {
+		o(&options)
 	}
 
-	return service, nil
-}
+	// if no TTL has been set, set it to DefaultTTL
+	if options.TTL.Seconds() == 0.0 {
+		options.TTL = DefaultTTL
+	}
 
-func (m *Registry) ListServices() ([]*registry.Service, error) {
-	var services []*registry.Service
-	m.RLock()
-	for _, service := range m.Services {
-		services = append(services, service...)
+	r := serviceToRecord(s)
+
+	if _, ok := m.records[s.Name]; !ok {
+		m.records[s.Name] = make(map[string]*record)
 	}
-	m.RUnlock()
-	return services, nil
+
+	if _, ok := m.records[s.Name][s.Version]; !ok {
+		m.records[s.Name][s.Version] = r
+		log.Debugf("Registry added new service: %s, version: %s", s.Name, s.Version)
+		go m.sendEvent(&registry.Result{Action: "update", Service: s})
+		return nil
+	}
+
+	addedNodes := false
+	for _, n := range s.Nodes {
+		if _, ok := m.records[s.Name][s.Version].Nodes[n.Id]; !ok {
+			addedNodes = true
+			metadata := make(map[string]string)
+			for k, v := range n.Metadata {
+				metadata[k] = v
+				m.records[s.Name][s.Version].Nodes[n.Id] = &node{
+					Node: &registry.Node{
+						Id:       n.Id,
+						Address:  n.Address,
+						Metadata: metadata,
+					},
+					TTL:      options.TTL,
+					LastSeen: time.Now(),
+				}
+			}
+		}
+	}
+
+	if addedNodes {
+		log.Debugf("Registry added new node to service: %s, version: %s", s.Name, s.Version)
+		go m.sendEvent(&registry.Result{Action: "update", Service: s})
+		return nil
+	}
+
+	// refresh TTL and timestamp
+	for _, n := range s.Nodes {
+		log.Debugf("Updated registration for service: %s, version: %s", s.Name, s.Version)
+		m.records[s.Name][s.Version].Nodes[n.Id].TTL = options.TTL
+		m.records[s.Name][s.Version].Nodes[n.Id].LastSeen = time.Now()
+	}
+
+	return nil
 }
 
-func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption) error {
+func (m *Registry) Deregister(s *registry.Service) error {
 	m.Lock()
 	defer m.Unlock()
 
-	if service, ok := m.Services[s.Name]; !ok {
-		m.Services[s.Name] = []*registry.Service{s}
-		go m.sendEvent(&registry.Result{Action: "update", Service: s})
-	} else {
-		svcCount := len(service)
-		svcNodeCounts := make(map[string]map[string]int)
-		for _, s := range service {
-			if _, ok := svcNodeCounts[s.Name]; !ok {
-				svcNodeCounts[s.Name] = make(map[string]int)
+	if _, ok := m.records[s.Name]; ok {
```

Comment:
Should we do if !ok to return immediately so the rest of this is not nested under an if statement?

Reply:
Nope, because in the nested `if` we gradually delete records from the map (and ths return memory to heap) if they're no longer needed:
1. node
2. service version record
3. the whole service record